{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os\n",
    "import json\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO, AlignIO, Align, SeqRecord\n",
    "from collections import Counter\n",
    "import dendropy\n",
    "from dendropy.calculate import treecompare\n",
    "from Bio import Phylo\n",
    "import subprocess\n",
    "from subprocess import PIPE, Popen\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def noise_filter(column):\n",
    "    \"\"\"\n",
    "    noise_filter is a function that determines whether an alignment column is noisy or not.\n",
    "    :param column: multiple sequence alignment column.\n",
    "    :return: boolean.\n",
    "    \"\"\"\n",
    "    indel = '-'\n",
    "    amino_counter = Counter(column)\n",
    "    if column.count(indel) > len(column) / 2:\n",
    "        return True\n",
    "    if len([key for key, value in amino_counter.items() if value == 1 ]) >= len(column) / 2:\n",
    "        return True\n",
    "    if max(amino_counter.values()) <= 2:\n",
    "        return True\n",
    "    return False     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reduce_noise(multiple_seq_alignment):\n",
    "    \"\"\"\n",
    "    reduce_noise is a function that filters all noisy columns using the function noise_filter.\n",
    "    :param multiple_seq_alignment: MultipleSeqAlignment.\n",
    "    :return: a MultipleSeqAlignment.\n",
    "    \"\"\"\n",
    "    noise_free_column_list = []\n",
    "    new_aligned_list = []\n",
    "    for i in range(multiple_seq_alignment.get_alignment_length()):\n",
    "        if not noise_filter(multiple_seq_alignment[:, i]):\n",
    "            noise_free_column_list.append(multiple_seq_alignment[:, i])\n",
    "    for i in range(len(multiple_seq_alignment)):\n",
    "        new_aligned_list.append(''.join([x[i] for x in noise_free_column_list]))\n",
    "    record_list = Align.MultipleSeqAlignment([\n",
    "        SeqRecord.SeqRecord(Seq(new_seq), id=record.id, name=record.name, description=record.description)\n",
    "        for new_seq, record in zip(new_aligned_list, multiple_seq_alignment)\n",
    "     ])\n",
    "    return record_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perform_noise_reduction(alignment_path):\n",
    "    \"\"\"\n",
    "    perform_noise_reduction is a function that takes the parameter alignment path and applies \n",
    "    the reduce_noise function.\n",
    "    :param alignment_path: msa path.\n",
    "    :return: a MultipleSeqAlignment.\n",
    "    \"\"\"\n",
    "    with open(alignment_path, mode='r') as aligned_file:\n",
    "        my_sequence_recorded = AlignIO.read(aligned_file, 'fasta')\n",
    "    return reduce_noise(my_sequence_recorded)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def computing_and_writing_alignment_tree(msa_filename, tree_outfile):\n",
    "    \"\"\"\n",
    "    computing_and_writing_alignment_tree is a function that pipes two processes,\n",
    "    it first creates a distance matrix for each msa using fastprot, \n",
    "    an then infeers an alignment tree using fnj. The function writes the inferred \n",
    "    tree in a file per msa.\n",
    "    :param msa_filename: msa path.\n",
    "    :param tree_outfile: infered tree path.\n",
    "    :return: file containing the inferred tree.\n",
    "    \"\"\"\n",
    "    args= \"cat \" + msa_filename + \" | fastprot | fnj -O newick -o \" + tree_outfile\n",
    "    child_noise_reduced =subprocess.Popen(args, shell=True)\n",
    "    child_noise_reduced.wait()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_trees(ref_tree, infered_tree, is_bipartitions_updated=False):\n",
    "     \"\"\"\n",
    "    compare_trees is a function that computes the simmetric difference between \n",
    "    the reference tree and the inferred tree.\n",
    "    :param ref_tree: reference tree read using dendropy.\n",
    "    :param infered_tree: inferred tree read using dendropy.\n",
    "    :param is_bipartitions_updated: recalculates bipartitions.\n",
    "    :return: (int) symmetric distanc between the reference and infered tree.\n",
    "    \"\"\"\n",
    "    sd = treecompare.symmetric_difference(ref_tree, infered_tree)\n",
    "    return (sd)\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_distance_between_trees(infered_tree_file_path, reference_tree):\n",
    "    \"\"\"\n",
    "    compute_distance_between_trees is a function that reads inferred trees using \n",
    "    dendropy and computes the distance using the function compare_trees\n",
    "    :param infered_tree_file_path: path\n",
    "    :param reference_tree: reference tree read using dendropy.\n",
    "    :return: (int) symmetric distanc between the reference and infered tree.\n",
    "    \"\"\"\n",
    "    with open(infered_tree_file_path, mode='r') as reduced_noise_tree_file:\n",
    "            reduced_noise_tree_str = ''.join(list(reduced_noise_tree_file)) \n",
    "            reduced_noise_tree = dendropy.Tree.get_from_string(\n",
    "                reduced_noise_tree_str,\n",
    "                schema=\"newick\",\n",
    "                taxon_namespace=tns\n",
    "            )\n",
    "    return compare_trees(reference_tree, reduced_noise_tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "This script will perform all the program's required tasks. \n",
    "It will first create a list of all the data directories where we will reduce noise, \n",
    "new directories for storing the reduced allignments and an empty dictionary \n",
    "with key: main directory name (e.g asymmetric_0.5) keyvalue: distances, which will \n",
    "later contained a nested dictionary with key: filename andkeyvalues: list with two \n",
    "elements, the first one: distance of the reference tree with the original alignement\n",
    "and the second one: distance of the ref tree with the noise reduced alignment.\n",
    "\n",
    "The program will start parsing through the raw data, it will first read the reference tree,\n",
    "set the keys of the dictionary and write the noise reduced alignment file using the\n",
    "perform_noise_reduction function, afterwards it will generate infered tree files, for both:\n",
    "the raw data and the noise reduced data using the function computing_and_writing_alignment_tree\n",
    "lastly it will compute the distance between the reference tree and both inferred tree cases \n",
    "(with an without noise reduction). The filenames (key) and distances (keyvalues) will be nested\n",
    "in the dictionary.\n",
    "\n",
    "The program will write a json file in the results folder containing the dictionary.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/asymmetric_1.0.tree',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s001.align.1.msl',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s001.align.1.tree',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s002.align.1.msl',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s002.align.1.tree',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s003.align.1.msl',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s003.align.1.tree',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s004.align.1.msl',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s004.align.1.tree',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s005.align.1.msl',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s005.align.1.tree',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s006.align.1.msl',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s006.align.1.tree',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s007.align.1.msl',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s007.align.1.tree',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s008.align.1.msl',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s008.align.1.tree',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s009.align.1.msl',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s009.align.1.tree',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s010.align.1.msl',\n",
       " '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data/test_2/s010.align.1.tree']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glob.glob(original_dir_path +'/*')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_dir = '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/test_data'\n",
    "reduced_dir = '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/data/reduced_test_data'\n",
    "result_dir =  '/Users/user/Desktop/MSM/03-Applied_Bioinformatics/Labs/03_Project/protein_multial_noise_reduction/results'\n",
    "directories = []\n",
    "tns = dendropy.TaxonNamespace()\n",
    "\n",
    "for folder in glob.glob(original_dir +'/*'):\n",
    "    sub_folder_name = folder.split('/')[-1]\n",
    "    directories.append(sub_folder_name)\n",
    "\n",
    "\n",
    "# this will contain all the symmetric distances between infered trees \n",
    "# and reference trees, for all different data subdirectories\n",
    "compare_trees_dictionary = dict()\n",
    "\n",
    "for directory in directories: # directories = [0.5Alignment, 1.0Alignment, ...]\n",
    "    \n",
    "    # create *0.5Alignment, *1.0Alignment, ... subdirectories into the reduced_data directory\n",
    "    # and getting the reference tree for each subdirectory of alignments.\n",
    "    new_folder_path = os.path.join(reduced_dir, directory)\n",
    "    os.makedirs(new_folder_path, exist_ok=True)\n",
    "    original_dir_path = os.path.join(original_dir, directory)\n",
    "    reference_tree_path = glob.glob(original_dir_path +'/*.tree')[0]\n",
    "    with open(reference_tree_path, mode='r') as ref_tree_file:\n",
    "        ref_tree_str = ''.join(list(ref_tree_file))\n",
    "        reference_tree = dendropy.Tree.get_from_string(\n",
    "        ref_tree_str,\n",
    "        schema=\"newick\",\n",
    "        taxon_namespace=tns\n",
    "    )\n",
    "    folder_dict_key_name = original_dir_path.split('/')[-1]\n",
    "    \n",
    "    compare_trees_dictionary[folder_dict_key_name] = dict()\n",
    "    \n",
    "    for alignment_path in glob.glob(original_dir_path + '/*.msl'):\n",
    "        # performing noise reduction and writing the alignment\n",
    "        # into the corresponding noise_reduction directory\n",
    "        record_list = perform_noise_reduction(alignment_path)\n",
    "        reduced_alignment_name = alignment_path.split('/')[-1]\n",
    "        reduced_filename_out = os.path.join(new_folder_path, reduced_alignment_name)\n",
    "        AlignIO.write(record_list, reduced_filename_out, \"fasta\")\n",
    "        \n",
    "        # computing and writing noise reduced alignment trees\n",
    "        tree_outfile_reduced = reduced_filename_out[:-3] + 'tree'\n",
    "        computing_and_writing_alignment_tree(reduced_filename_out, tree_outfile_reduced)\n",
    "        \n",
    "        # computing and writing original alignment trees\n",
    "        alignment_name = alignment_path.split('/')[-1]\n",
    "        filename_out = os.path.join(original_dir_path, alignment_name)\n",
    "        tree_outfile = filename_out[:-3] + 'tree'\n",
    "        computing_and_writing_alignment_tree(alignment_path, tree_outfile)\n",
    "        \n",
    "        # computing distance between ref tree and infered trees\n",
    "        noise_reduced_distance = compute_distance_between_trees(tree_outfile_reduced, reference_tree)\n",
    "        original_distance = compute_distance_between_trees(tree_outfile, reference_tree)\n",
    "        \n",
    "        alignment_key_name = alignment_name[:-3]\n",
    "        compare_trees_dictionary[folder_dict_key_name][alignment_key_name] = (\n",
    "            original_distance, noise_reduced_distance\n",
    "        )\n",
    "\n",
    "distance_results_path = os.path.join(result_dir, 'distance_result_dict')\n",
    "with open(distance_results_path, 'w') as result_dir_file:\n",
    "    json.dump(compare_trees_dictionary, result_dir_file)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
